format-avro/src/main/java/io/cdap/plugin/format/avro/output/DelegatingAvroKeyRecordWriter.java

```diff
@@ -0,0 +1,95 @@
+package io.cdap.plugin.format.avro.output;
+
+import org.apache.avro.file.CodecFactory;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.hadoop.io.AvroSerialization;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapreduce.AvroKeyRecordWriter;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Record Writer for Avro records that delegates to additional record writers based on the Schema hash.
+ */
+public class DelegatingAvroKeyRecordWriter extends RecordWriter<AvroKey<GenericRecord>, NullWritable> {
+  private final Map<Integer, RecordWriter<AvroKey<GenericRecord>, NullWritable>> delegateMap;
+  private final TaskAttemptContext context;
+  private final CodecFactory codecFactory;
+  private final Function<TaskAttemptContext, OutputStream> outputStreamSupplier;
+  private final int syncInterval;
+
+  public DelegatingAvroKeyRecordWriter(TaskAttemptContext context,
+                                       CodecFactory codecFactory,
+                                       Function<TaskAttemptContext, OutputStream> outputStreamSupplier,
+                                       int syncInterval) {
+    super();
+    this.delegateMap = new HashMap<>();
+    this.context = context;
+    this.codecFactory = codecFactory;
+    this.outputStreamSupplier = outputStreamSupplier;
+    this.syncInterval = syncInterval;
+  }
+
+  @Override
+  public void write(AvroKey<GenericRecord> key, NullWritable value) throws IOException, InterruptedException {
+    RecordWriter<AvroKey<GenericRecord>, NullWritable> delegate =
+      delegateMap.computeIfAbsent(key.datum().getSchema().hashCode(), (k) -> {
+        try {
+          Configuration conf = context.getConfiguration();
+          GenericData dataModel = AvroSerialization.createDataModel(conf);
+
+          return new AvroKeyRecordWriter<>(key.datum().getSchema(),
+                                           dataModel,
+                                           codecFactory,
+                                           outputStreamSupplier.apply(context),
+                                           syncInterval);
+        } catch (IOException ioe) {
+          throw new RuntimeException("Unable to initialize Avro Record Writer", ioe);
+        }
+      });
+
+    delegate.write(key, value);
+  }
+
+  @Override
+  public void close(TaskAttemptContext context)  {
+    RuntimeException ex = new RuntimeException("Unable to close delegate.");
+
+    for (RecordWriter<AvroKey<GenericRecord>, NullWritable> delegate : delegateMap.values()) {
+      try {
+        delegate.close(context);
+      } catch (IOException | InterruptedException e) {
```

Comment:
don't need to catch the InterruptedException, better to stop doing all work and let it propagate up immediately